### PR TITLE
Session changes

### PR DESCRIPTION
**Pull request by 23TNC**
Writing session variables to .powervs in home directory so they are persistant between GetSession calls

Signed-off-by: wolf <cory.klokman@ibm.com>

### CODE REVIEW

File: pkg/asset/installconfig/powervs/session.go [hunk 1]
```diff
@@ -19,105 +24,208 @@ var (
...
+	// First grab variables from the installer written authFilePath
+	logrus.Debug("Gathering variables from AuthFile")
+	err = getPISessionVarsFromAuthFile(&pisv)
+	if err != nil {
+		return nil, err
+	}
+
+	// Second grab variables from files
+
+	// .bluemix/config.json doesn't seem to hold anything useful but other files might
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go]: let's remove this comment just b/c it's a bit too wordy.
but on the topic of the ~/.bluemix/config.json, it'll only contain something if a user intentionally created it and gave it that name (`ibmcloud iam api-key-create test --file ~/.bluemix/apikey_test.json --output json`).

Comment (clnperez) [reply pkg/asset/installconfig/powervs/session.go]: what do you think about just removing any checking for this file? it's not documented anywhere and i don't even remember where I found it when this project started. imo it's just a complication, and we can document that for IPI, users have to use the env vars.

**Review by clnperez — CHANGES_REQUESTED** [review]
just one suggestion

File: pkg/asset/installconfig/powervs/session.go [hunk 2]
```diff
@@ -19,105 +24,202 @@ var (
...
+	var pisv PISessionVars
+
+	// First grab variables from the installer written authFilePath
+	logrus.Debug("Gathering variables from AuthFile")
+	err = getPISessionVarsFromAuthFile(&pisv)
+	if err != nil {
+		return nil, err
+	}
+
+	// Second grab variables from files, currently there are no other files to check
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go]: can you remove `currently there are no other files to check`? Without all the historical context we have this doesn't make sense

File: pkg/asset/installconfig/powervs/session.go [hunk 3]
```diff
@@ -19,105 +24,202 @@ var (
...
-
 	return &Session{Session: s}, nil
 }
 
-/*
-//  https://github.com/IBM-Cloud/power-go-client/blob/master/ibmpisession/ibmpowersession.go
-*/
+var (
+	defaultAuthFilePath    = filepath.Join(os.Getenv("HOME"), ".powervs", "config.json")
+	defaultBluemixFilePath = filepath.Join(os.Getenv("HOME"), ".bluemix", "config.json")
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go]: also remove ^

**Review by clnperez — CHANGES_REQUESTED** [review]

File: pkg/asset/installconfig/powervs/session.go (hunk 3 above)
Comment (23TNC) [reply pkg/asset/installconfig/powervs/session.go]: Done

File: pkg/asset/installconfig/powervs/session.go [hunk 4]
```diff
@@ -19,105 +24,201 @@ var (
...
+		return err
 	}
 
-	return s, err
+	err = os.MkdirAll(filepath.Dir(authFilePath), 0700)
+	if err != nil {
+		return err
+	}
+	logrus.Debug("Saving variables to ", authFilePath)
+	return ioutil.WriteFile(authFilePath, jsonVars, 0600)
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:228]: can you add an error check for the WriteFile here?

File: pkg/asset/installconfig/powervs/session.go (hunk 2 above)
Comment (23TNC) [reply pkg/asset/installconfig/powervs/session.go]: Just tried to retain the logic here, removed it as per request.

File: pkg/asset/installconfig/powervs/session.go [hunk 5]
```diff
@@ -19,105 +24,201 @@ var (
...
-	apikey = getEnv(APIKeyEnvVars)
+	}
 
-	if len(apikey) == 0 {
+	if len(pisv.APIKey) == 0 {
 		err = survey.Ask([]*survey.Question{
 			{
 				Prompt: &survey.Password{
 					Message: "IBM Cloud API Key",
 					Help:    "The api key installation.\nhttps://cloud.ibm.com/iam/apikeys",
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go]: while you're making changes here, can you fix the capitalization in this string? api => API

File: pkg/asset/installconfig/powervs/session.go [hunk 6]
```diff
@@ -19,105 +24,201 @@ var (
...
+	logrus.Debug("Gathering variables from AuthFile")
+	err = getPISessionVarsFromAuthFile(&pisv)
+	if err != nil {
+		return nil, err
+	}
+
+	// Second grab variables from files, currently there are no other files to check
+
+	// Third grab variables from the users enviornment
+	logrus.Debug("Gathering variables from user enviornment")
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go]: /s/enviornment/environment

File: pkg/asset/installconfig/powervs/session.go [hunk 7]
```diff
@@ -19,105 +24,201 @@ var (
...
+		return nil, err
+	}
+
+	// Save variables to disk
+	err = savePISessionVars(&pisv)
+	if err != nil {
+		return nil, err
+	}
+
+	logrus.Debug("Done")
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go]: let's remove this or make it more descriptive. you could output a generic Done, but then still return an error below

File: pkg/asset/installconfig/powervs/session.go [hunk 8]
```diff
@@ -19,105 +24,201 @@ var (
...
+
+	// This is needed by ibmcloud code to gather dns information later
+	os.Setenv("IC_API_KEY", pisv.APIKey)
+	// We are using the iamtoken field to hold the api key
+	iamtoken := pisv.APIKey
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go]: can you remove this var and just pass `pisv.APIKey` straight into the func below? and this comment can go too.

File: pkg/asset/installconfig/powervs/session.go [hunk 9]
```diff
@@ -19,105 +24,201 @@ var (
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getPISessionVarsFromEnv(pisv *PISessionVars) error {
+
+	if len(pisv.ID) == 0 {
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:133]: put in a quick check to make sure this pointer isn't nil

**Review by clnperez — CHANGES_REQUESTED** [review]
